File: csharp_dotnetcore/Dialogs/ContactDialog.cs
using FoodBot.Services;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace FoodBot.Dialogs
{
    public class ContactDialog : ComponentDialog
    {
        public ContactDialog(string dialogId) : base(dialogId)
        {
            const string FOODBANKNAME = "foodBankName";
            const string EMAILADDRESS = "emailAddress";
            const string MESSAGE = "message";

            // ID of the child dialog that should be started anytime the component is started.
            this.InitialDialogId = dialogId;

            this.AddDialog(new ChoicePrompt("choicePrompt"));
            this.AddDialog(new TextPrompt("textPrompt"));
            this.AddDialog(new ConfirmPrompt("confirmPrompt"));

            // Define the conversation flow using the waterfall model.
            this.AddDialog(
                new WaterfallDialog(dialogId, new WaterfallStep[]
                    {
                        async (stepContext, ct) =>
                        {
                            return await stepContext.PromptAsync(
                                "choicePrompt",
                                new PromptOptions
                                {
                                    Choices = ChoiceFactory.ToChoices(ScheduleHelpers.GetFoodBanks()),
                                    Prompt = MessageFactory.Text("Which food bank would you like to contact?"),
                                    RetryPrompt= MessageFactory.Text("That's not a valid food bank! Please choose a valid food bank.")
                                },
                                ct
                            ).ConfigureAwait(false);;
                        },
                        async (stepContext, ct) =>
                        {
                            // Persist the food bank name for later waterfall steps to be able to access it
                            stepContext.Values.Add(FOODBANKNAME, ((FoundChoice)stepContext.Result).Value);

                            return await stepContext.PromptAsync("textPrompt", new PromptOptions
                                {
                                    Prompt = MessageFactory.Text($"Please enter an email address where {stepContext.Values[FOODBANKNAME]} can message you back at:")
                                },
                                ct
                            ).ConfigureAwait(false);;
                        },
                        async (stepContext, ct) =>
                        {
                            // Persist the email address for later waterfall steps to be able to access it
                            stepContext.Values.Add(EMAILADDRESS, (string)stepContext.Result);

                            return await stepContext.PromptAsync("textPrompt", new PromptOptions
                                {
                                    Prompt = MessageFactory.Text($"Please enter the message you'd like to send to {stepContext.Values[FOODBANKNAME]}:")
                                },
                                ct
                            ).ConfigureAwait(false);;
                        },
                        async (stepContext, ct) =>
                        {
                            // Persist the message for later waterfall steps to be able to access it
                            stepContext.Values.Add(MESSAGE, (string)stepContext.Result);

                            return await stepContext.PromptAsync("confirmPrompt", new PromptOptions
                                {
                                    Prompt = MessageFactory.Text($"Are you sure you'd like to send:\r \"{stepContext.Values[MESSAGE]}\"\r to {stepContext.Values[FOODBANKNAME]}?")
                                },
                                ct
                            ).ConfigureAwait(false);
                        },
                        async (stepContext, ct) =>
                        {
                            // Simulates sending of a message (nothing is actually being sent anywhere)
                            if ((bool)stepContext.Result)
                            {
                                ScheduleHelpers.SendFoodbankMessage((string)stepContext.Values[FOODBANKNAME], (string)stepContext.Values[EMAILADDRESS], (string)stepContext.Values[MESSAGE]);
                                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Great! We've sent your message to {stepContext.Values[FOODBANKNAME]}. Expect your response to be sent to {stepContext.Values[EMAILADDRESS]}"), ct).ConfigureAwait(false);;
                            }
                            else
                            {
                                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"No worries! I won't send your message"), ct).ConfigureAwait(false);
                            }

                            return await stepContext.EndDialogAsync().ConfigureAwait(false);;
                        }
                    }
                )
            );
        }
    }
}

File: javascript_nodejs/dialogs/ContactDialog.js
const { ComponentDialog, ConfirmPrompt, ChoicePrompt, TextPrompt, WaterfallDialog } = require('botbuilder-dialogs');
const { getFoodBanks, sendFoodBankMessage } = require('../services/schedule-helpers');

class ContactDialog extends ComponentDialog {
    constructor(dialogId) {
        super(dialogId);

        // ID of the child dialog that should be started anytime the component is started.
        this.initialDialogId = dialogId;

        // Define the prompts used in this conversation flow.
        this.addDialog(new ChoicePrompt('choicePrompt'));
        this.addDialog(new TextPrompt('textPrompt'));
        this.addDialog(new ConfirmPrompt('confirmPrompt'));

        // Define the conversation flow using a waterfall model.
        this.addDialog(new WaterfallDialog(dialogId, [
            async function (step) {
                const foodBanks = getFoodBanks();

                return await step.prompt('choicePrompt', {
                    choices: foodBanks,
                    prompt: "Which food bank would you like to contact?",
                    retryPrompt: "That's not a valid food bank! Please choose a valid food bank."
                });
            },
            async function (step) {
                // Persist the food bank name for later waterfall steps to be able to access it
                step.values.foodBankName = step.result.value;

                return await step.prompt('textPrompt', {
                    prompt: `Please enter an email address where ${step.values.foodBankName} can message you back at:`
                });
            },
            async function (step) {
                // Persist the email address for later waterfall steps to be able to access it
                step.values.email = step.result;

                return await step.prompt('textPrompt',
                    `Please enter the message you'd like to send to ${step.values.foodBankName}:`,
                );
            },
            async function (step) {
                // Persist the message for later waterfall steps to be able to access it
                step.values.message = step.result;

                return await step.prompt('confirmPrompt', {
                    prompt: `Are you sure you'd like to send:\r
                    "${step.values.message}"\r
                    to ${step.values.foodBankName}?`,
                })
            },
            async function (step) {
                //Simulates sending of a message (nothing is actually being sent anywhere)
                if (step.result) {
                    sendFoodBankMessage(step.result.foodBankName, step.result.message, step.result.email);
                    return await step.context.sendActivity(`Great! We've sent your message to ${step.values.foodBankName}. Expect your response to be sent to ${step.values.email}`);
                } else {
                    return await step.context.sendActivity(`No worries! I won't send your message`);
                }
            }
        ]));
    }
}

exports.ContactDialog = ContactDialog;